osm2lanes/src/algorithm.rs

```diff
@@ -1,610 +1,238 @@
+            (false, false) => continue,
+        };
 
-    if tags.is("sidewalk", "both") {
-        fwd_side.push(fwd(LaneType::Sidewalk));
-        back_side.push(back(LaneType::Sidewalk));
-    } else if tags.is("sidewalk", "separate") && cfg.inferred_sidewalks {
-        // TODO Need to snap separate sidewalks to ways. Until then, just do this.
-        fwd_side.push(fwd(LaneType::Sidewalk));
-        if !back_side.is_empty() {
-            back_side.push(back(LaneType::Sidewalk));
-        }
-    } else if tags.is("sidewalk", "right") {
-        if cfg.driving_side == DrivingSide::Right {
-            fwd_side.push(fwd(LaneType::Sidewalk));
+        let lane_type = if main || rank.lane_type != LaneType::Footway {
+            rank.lane_type
         } else {
-            back_side.push(back(LaneType::Sidewalk));
-        }
-    } else if tags.is("sidewalk", "left") {
-        if cfg.driving_side == DrivingSide::Right {
-            back_side.push(back(LaneType::Sidewalk));
+            LaneType::Sidewalk
+        };
+
+        let turns = if access_forward {
+            &t.forward
         } else {
-            fwd_side.push(fwd(LaneType::Sidewalk));
+            &t.backward
         }
-    }
+        .turn
+        .get(rank.main)
+        .and_then(|v| v.base())
+        .map(TurnDirection::from_muv)
+        .unwrap_or_default();
 
-    let mut need_fwd_shoulder = fwd_side
-        .last()
-        .map(|spec| spec.lt != LaneType::Sidewalk)
-        .unwrap_or(true);
-    let mut need_back_shoulder = back_side
-        .last()
-        .map(|spec| spec.lt != LaneType::Sidewalk)
-        .unwrap_or(true);
-    if tags.is_any(
-        osm::HIGHWAY,
-        vec!["motorway", "motorway_link", "construction"],
-    ) || tags.is("foot", "no")
-        || tags.is("access", "no")
-        || tags.is("motorroad", "yes")
-    {
-        need_fwd_shoulder = false;
-        need_back_shoulder = false;
-    }
-    // If it's a one-way, fine to not have sidewalks on both sides.
-    if tags.is("oneway", "yes") {
-        need_back_shoulder = false;
-    }
-
-    // If it's a non-motorized road and we already have some walkable lane, we don't need
-    // shoulders
-    if fwd_side
-        .iter()
-        .chain(back_side.iter())
-        .any(|spec| spec.lt.is_walkable())
-    {
-        need_fwd_shoulder = false;
-        need_back_shoulder = false;
-    }
-
-    // For living streets in Krakow, there aren't separate footways. People can walk in the street.
-    // For now, model that by putting shoulders.
-    if cfg.inferred_sidewalks || tags.is(osm::HIGHWAY, "living_street") {
-        if need_fwd_shoulder {
-            fwd_side.push(fwd(LaneType::Shoulder));
-        }
-        if need_back_shoulder {
-            back_side.push(back(LaneType::Shoulder));
-        }
+        return (lane_type, dir, turns);
     }
+    (LaneType::Construction, Direction::Fwd, EnumSet::new())
```

Comment:
There's also a lifecycle check elsewhere that assigns construction. In this fall-back case, is construction the only case in OSM, or is this osm2streets not yet modeling some situation? No rank has access, so it might be something like a military/delivery-only access?

Reply:
This currently just imitates the old algorithm which would fall back to `Construction` if the way is not accessable. Maybe it's worth having a separate `LaneType` for those cases.